File: HexaFlipTests/Source/UI/Screens/JCSFlipUIPlayerMenuScreenTest.m
//
//  JCSFlipUIPlayerMenuScreenTest.m
//  HexaFlip
//
//  Created by Christian Schuster on 09.02.14.
//  Copyright (c) 2014 Christian Schuster. All rights reserved.
//

#import "JCSButton.h"
#import "JCSRadioMenu.h"
#import "JCSFlipUIPlayerMenuScreen.h"
#import "JCSFlipUIEvents.h"
#import "JCSFlipPlayerLocal.h"
#import "JCSFlipPlayerAI.h"
#import "JCSFlipGameState.h"

#import "OCMock.h"

@interface JCSFlipUIPlayerMenuScreenTest : XCTestCase
@end

@implementation JCSFlipUIPlayerMenuScreenTest {
    JCSFlipUIPlayerMenuScreen *_underTest;
    id _observerMock;
    id _gameStateMock;
}

- (void)setUp {
    [super setUp];
    _underTest = [[JCSFlipUIPlayerMenuScreen alloc] init];
    _observerMock = [OCMockObject observerMock];
    _gameStateMock = [OCMockObject mockForClass:[JCSFlipGameState class]];
    NSNotificationCenter *nc = [NSNotificationCenter defaultCenter];
    [nc addMockObserver:_observerMock name:JCS_FLIP_UI_BACK_EVENT_NAME object:nil];
    [nc addMockObserver:_observerMock name:JCS_FLIP_UI_PREPARE_GAME_EVENT_NAME object:nil];
    [nc addMockObserver:_observerMock name:JCS_FLIP_UI_PLAY_GAME_EVENT_NAME object:nil];
}

- (void)tearDown {
    NSNotificationCenter *nc = [NSNotificationCenter defaultCenter];
    [nc removeObserver:_observerMock];
    [super tearDown];
}

- (JCSButton *)backButtonForScreen:(JCSFlipUIPlayerMenuScreen *)screen {
    CCMenu *menu = [screen.children objectAtIndex:0];
    return [menu.children objectAtIndex:0];
}

- (JCSButton *)playButtonForScreen:(JCSFlipUIPlayerMenuScreen *)screen {
    CCMenu *menu = [screen.children objectAtIndex:0];
    return [menu.children objectAtIndex:1];
}

- (JCSButton *)playerALocalButtonForScreen:(JCSFlipUIPlayerMenuScreen *)screen {
    JCSRadioMenu *radioPlayerA = [screen.children objectAtIndex:1];
    return [radioPlayerA.children objectAtIndex:0];
}

- (JCSButton *)playerAAIEasyButtonForScreen:(JCSFlipUIPlayerMenuScreen *)screen {
    JCSRadioMenu *radioPlayerA = [screen.children objectAtIndex:1];
    return [radioPlayerA.children objectAtIndex:1];
}

- (JCSButton *)playerAAIMediumButtonForScreen:(JCSFlipUIPlayerMenuScreen *)screen {
    JCSRadioMenu *radioPlayerA = [screen.children objectAtIndex:1];
    return [radioPlayerA.children objectAtIndex:2];
}

- (JCSButton *)playerAAIHardButtonForScreen:(JCSFlipUIPlayerMenuScreen *)screen {
    JCSRadioMenu *radioPlayerA = [screen.children objectAtIndex:1];
    return [radioPlayerA.children objectAtIndex:3];
}

- (JCSButton *)playerBLocalButtonForScreen:(JCSFlipUIPlayerMenuScreen *)screen {
    JCSRadioMenu *radioPlayerB = [screen.children objectAtIndex:2];
    return [radioPlayerB.children objectAtIndex:0];
}

- (JCSButton *)playerBAIEasyButtonForScreen:(JCSFlipUIPlayerMenuScreen *)screen {
    JCSRadioMenu *radioPlayerB = [screen.children objectAtIndex:2];
    return [radioPlayerB.children objectAtIndex:1];
}

- (JCSButton *)playerBAIMediumButtonForScreen:(JCSFlipUIPlayerMenuScreen *)screen {
    JCSRadioMenu *radioPlayerB = [screen.children objectAtIndex:2];
    return [radioPlayerB.children objectAtIndex:2];
}

- (JCSButton *)playerBAIHardButtonForScreen:(JCSFlipUIPlayerMenuScreen *)screen {
    JCSRadioMenu *radioPlayerB = [screen.children objectAtIndex:2];
    return [radioPlayerB.children objectAtIndex:3];
}

- (void)testEventTriggeredWhenBackButtonActivated {
    [[_observerMock expect] notificationWithName:JCS_FLIP_UI_BACK_EVENT_NAME object:_underTest];
    
    [[self backButtonForScreen:_underTest] activate];
    
    [_observerMock verify];
}

- (void)checkEventsWhenActivatingButton:(JCSButton *)button checkPlayerDescriptionsUsingBlock:(void(^)(NSString *playerADescription, NSString *playerBDescription))block {
    __block NSDictionary *capturedUserInfo;
    [[_observerMock expect] notificationWithName:JCS_FLIP_UI_PREPARE_GAME_EVENT_NAME object:_underTest userInfo:[OCMArg checkWithBlock:^BOOL(NSDictionary *userInfo) {
        capturedUserInfo = userInfo;
        return YES;
    }]];
    [[_observerMock expect] notificationWithName:JCS_FLIP_UI_PLAY_GAME_EVENT_NAME object:_underTest];
    
    [[[_gameStateMock expect] andReturn:_gameStateMock] alloc];
    (void) [[[_gameStateMock expect] andReturn:_gameStateMock] initWithSize:5 playerToMove:JCSFlipPlayerSideA];
    
    [button activate];
    [[self playButtonForScreen:_underTest] activate];
    
    [_observerMock verify];
    
    XCTAssertEqual(capturedUserInfo.count, 1u);
    JCSFlipUIPrepareGameEventData *capturedEventData = capturedUserInfo[capturedUserInfo.keyEnumerator.nextObject];
    XCTAssertNotNil(capturedEventData);
    XCTAssertEqual(capturedEventData.gameState, _gameStateMock);
    XCTAssertNil(capturedEventData.match);
    XCTAssertFalse(capturedEventData.animateLastMove);
    XCTAssertFalse(capturedEventData.moveInputDisabled);
    block(capturedEventData.playerA.description, capturedEventData.playerB.description);
}

static NSString *descriptionForLocalPlayer = @"(Local Player)";
static NSString *descriptionForAIEasyPlayer = @"(AI player; (NegaScout Algorithm; (Transposition Table; Size 1000000); (Possession Heuristic); Depth 1))";
static NSString *descriptionForAIMediumPlayer = @"(AI player; (NegaScout Algorithm; (Transposition Table; Size 1000000); (Possession Heuristic); Depth 4))";
static NSString *descriptionForAIHardPlayer = @"(AI player; (NegaScout Algorithm; (Transposition Table; Size 1000000); (Possession Heuristic); Depth 6))";

- (void)testEventsWhenPlayerTypesChangedAndPlayButtonActivated {
    [self checkEventsWhenActivatingButton:nil checkPlayerDescriptionsUsingBlock:^(NSString *playerADescription, NSString *playerBDescription) {
        XCTAssertEqualObjects(playerADescription, descriptionForLocalPlayer);
        XCTAssertEqualObjects(playerBDescription, descriptionForAIMediumPlayer);
    }];
    [self checkEventsWhenActivatingButton:[self playerAAIEasyButtonForScreen:_underTest] checkPlayerDescriptionsUsingBlock:^(NSString *playerADescription, NSString *playerBDescription) {
        XCTAssertEqualObjects(playerADescription, descriptionForAIEasyPlayer);
        XCTAssertEqualObjects(playerBDescription, descriptionForAIMediumPlayer);
    }];
    [self checkEventsWhenActivatingButton:[self playerBLocalButtonForScreen:_underTest] checkPlayerDescriptionsUsingBlock:^(NSString *playerADescription, NSString *playerBDescription) {
        XCTAssertEqualObjects(playerADescription, descriptionForAIEasyPlayer);
        XCTAssertEqualObjects(playerBDescription, descriptionForLocalPlayer);
    }];
    [self checkEventsWhenActivatingButton:[self playerAAIHardButtonForScreen:_underTest] checkPlayerDescriptionsUsingBlock:^(NSString *playerADescription, NSString *playerBDescription) {
        XCTAssertEqualObjects(playerADescription, descriptionForAIHardPlayer);
        XCTAssertEqualObjects(playerBDescription, descriptionForLocalPlayer);
    }];
    [self checkEventsWhenActivatingButton:[self playerBAIMediumButtonForScreen:_underTest] checkPlayerDescriptionsUsingBlock:^(NSString *playerADescription, NSString *playerBDescription) {
        XCTAssertEqualObjects(playerADescription, descriptionForAIHardPlayer);
        XCTAssertEqualObjects(playerBDescription, descriptionForAIMediumPlayer);
    }];
    [self checkEventsWhenActivatingButton:[self playerAAIMediumButtonForScreen:_underTest] checkPlayerDescriptionsUsingBlock:^(NSString *playerADescription, NSString *playerBDescription) {
        XCTAssertEqualObjects(playerADescription, descriptionForAIMediumPlayer);
        XCTAssertEqualObjects(playerBDescription, descriptionForAIMediumPlayer);
    }];
    [self checkEventsWhenActivatingButton:[self playerBAIEasyButtonForScreen:_underTest] checkPlayerDescriptionsUsingBlock:^(NSString *playerADescription, NSString *playerBDescription) {
        XCTAssertEqualObjects(playerADescription, descriptionForAIMediumPlayer);
        XCTAssertEqualObjects(playerBDescription, descriptionForAIEasyPlayer);
    }];
    [self checkEventsWhenActivatingButton:[self playerBAIHardButtonForScreen:_underTest] checkPlayerDescriptionsUsingBlock:^(NSString *playerADescription, NSString *playerBDescription) {
        XCTAssertEqualObjects(playerADescription, descriptionForAIMediumPlayer);
        XCTAssertEqualObjects(playerBDescription, descriptionForAIHardPlayer);
    }];
    [self checkEventsWhenActivatingButton:[self playerALocalButtonForScreen:_underTest] checkPlayerDescriptionsUsingBlock:^(NSString *playerADescription, NSString *playerBDescription) {
        XCTAssertEqualObjects(playerADescription, descriptionForLocalPlayer);
        XCTAssertEqualObjects(playerBDescription, descriptionForAIHardPlayer);
    }];
    
}

@end

File: HexaFlipTests/Source/UI/Screens/JCSFlipUIPlayerMenuScreenTest.swift
//
//  JCSFlipUIPlayerMenuScreenTest.swift
//  HexaFlip
//
//  Created by Christian Schuster on 05.11.14.
//  Copyright (c) 2014 Christian Schuster. All rights reserved.
//

import Foundation
import GameKit
import XCTest

class JCSFlipUIPlayerMenuScreenTest : XCTestCase {
    
    let underTest = JCSFlipUIPlayerMenuScreen()
    var events: [NSNotification] = []
    
    override func setUp() {
        NSNotificationCenter.defaultCenter().addObserver(self, selector: "eventTriggered:", name: nil, object: nil)
    }
    
    override func tearDown() {
        NSNotificationCenter.defaultCenter().removeObserver(self)
    }

    func testEventTriggeredWhenBackButtonActivated() {
        backButtonForScreen(underTest).activate()
        XCTAssertEqual(["JCS_FLIP_UI_BACK_EVENT_NAME"], events.map { notification in notification.name })
    }
    
    func testEventsWhenPlayerTypesChangedAndPlayButtonActivated() {
        let descriptionForLocalPlayer = "(Local Player)"
        let descriptionForAIEasyPlayer = "(AI player; (NegaScout Algorithm; (Transposition Table; Size 1000000); (Possession Heuristic); Depth 1))"
        let descriptionForAIMediumPlayer = "(AI player; (NegaScout Algorithm; (Transposition Table; Size 1000000); (Possession Heuristic); Depth 4))"
        let descriptionForAIHardPlayer = "(AI player; (NegaScout Algorithm; (Transposition Table; Size 1000000); (Possession Heuristic); Depth 6))"
        
        checkEventsWhenActivatingButton(nil, expectedPlayerADescription: descriptionForLocalPlayer, expectedPlayerBDescription: descriptionForAIMediumPlayer)
        checkEventsWhenActivatingButton(playerAAIEasyButtonForScreen(underTest), expectedPlayerADescription: descriptionForAIEasyPlayer, expectedPlayerBDescription: descriptionForAIMediumPlayer)
        checkEventsWhenActivatingButton(playerBLocalButtonForScreen(underTest), expectedPlayerADescription: descriptionForAIEasyPlayer, expectedPlayerBDescription: descriptionForLocalPlayer)
        checkEventsWhenActivatingButton(playerAAIHardButtonForScreen(underTest), expectedPlayerADescription: descriptionForAIHardPlayer, expectedPlayerBDescription: descriptionForLocalPlayer)
        checkEventsWhenActivatingButton(playerBAIMediumButtonForScreen(underTest), expectedPlayerADescription: descriptionForAIHardPlayer, expectedPlayerBDescription: descriptionForAIMediumPlayer)
        checkEventsWhenActivatingButton(playerAAIMediumButtonForScreen(underTest), expectedPlayerADescription: descriptionForAIMediumPlayer, expectedPlayerBDescription: descriptionForAIMediumPlayer)
        checkEventsWhenActivatingButton(playerBAIEasyButtonForScreen(underTest), expectedPlayerADescription: descriptionForAIMediumPlayer, expectedPlayerBDescription: descriptionForAIEasyPlayer)
        checkEventsWhenActivatingButton(playerBAIHardButtonForScreen(underTest), expectedPlayerADescription: descriptionForAIMediumPlayer, expectedPlayerBDescription: descriptionForAIHardPlayer)
        checkEventsWhenActivatingButton(playerALocalButtonForScreen(underTest), expectedPlayerADescription: descriptionForLocalPlayer, expectedPlayerBDescription: descriptionForAIHardPlayer)
    }

    
    func checkEventsWhenActivatingButton(button: JCSButton?, expectedPlayerADescription: String, expectedPlayerBDescription: String) {
        events.removeAll()
        
        button?.activate()

        XCTAssertEqual(events.count, 0)
        
        playButtonForScreen(underTest).activate()
        
        XCTAssertEqual(events.count, 2)
        
        let prepareEvent = events[0]
        XCTAssertEqual(prepareEvent.name, "JCS_FLIP_UI_PREPARE_GAME_EVENT_NAME")
        XCTAssertNotNil(prepareEvent.userInfo)
        XCTAssertEqual(prepareEvent.userInfo!.count, 1)
        let prepareEventData = prepareEvent.userInfo!.values.first! as JCSFlipUIPrepareGameEventData
        XCTAssertNotNil(prepareEventData)
        XCTAssertEqual(prepareEventData.gameState.playerToMove, JCSFlipPlayerSide.A)
        XCTAssertEqual(prepareEventData.gameState.cellCountPlayerA, 3)
        XCTAssertEqual(prepareEventData.gameState.cellCountPlayerB, 3)
        XCTAssertEqual(prepareEventData.gameState.cellCountEmpty, 54)
        XCTAssertNil(prepareEventData.match)
        XCTAssertFalse(prepareEventData.animateLastMove)
        XCTAssertFalse(prepareEventData.moveInputDisabled)
        XCTAssertEqual(prepareEventData.playerA.description, expectedPlayerADescription)
        XCTAssertEqual(prepareEventData.playerB.description, expectedPlayerBDescription)
        
        let playEvent = events[1]
        XCTAssertEqual(playEvent.name, "JCS_FLIP_UI_PLAY_GAME_EVENT_NAME")
        XCTAssertNil(playEvent.userInfo)
    }


    func descendantOfNode(node: CCNode, indexes: UInt...) -> CCNode {
        var cur = node
        for index in indexes {
            cur = cur.children.objectAtIndex(index) as CCNode
        }
        return cur
    }
    
    func backButtonForScreen(screen: JCSFlipUIPlayerMenuScreen) -> JCSButton {
        return descendantOfNode(screen, indexes: 0, 0) as JCSButton
    }

    func playButtonForScreen(screen: JCSFlipUIPlayerMenuScreen) -> JCSButton {
        return descendantOfNode(screen, indexes: 0, 1) as JCSButton
    }

    func playerALocalButtonForScreen(screen: JCSFlipUIPlayerMenuScreen) -> JCSButton {
        return descendantOfNode(screen, indexes: 1, 0) as JCSButton
    }

    func playerAAIEasyButtonForScreen(screen: JCSFlipUIPlayerMenuScreen) -> JCSButton {
        return descendantOfNode(screen, indexes: 1, 1) as JCSButton
    }

    func playerAAIMediumButtonForScreen(screen: JCSFlipUIPlayerMenuScreen) -> JCSButton {
        return descendantOfNode(screen, indexes: 1, 2) as JCSButton
    }
    
    func playerAAIHardButtonForScreen(screen: JCSFlipUIPlayerMenuScreen) -> JCSButton {
        return descendantOfNode(screen, indexes: 1, 3) as JCSButton
    }

    func playerBLocalButtonForScreen(screen: JCSFlipUIPlayerMenuScreen) -> JCSButton {
        return descendantOfNode(screen, indexes: 2, 0) as JCSButton
    }
    
    func playerBAIEasyButtonForScreen(screen: JCSFlipUIPlayerMenuScreen) -> JCSButton {
        return descendantOfNode(screen, indexes: 2, 1) as JCSButton
    }
    
    func playerBAIMediumButtonForScreen(screen: JCSFlipUIPlayerMenuScreen) -> JCSButton {
        return descendantOfNode(screen, indexes: 2, 2) as JCSButton
    }
    
    func playerBAIHardButtonForScreen(screen: JCSFlipUIPlayerMenuScreen) -> JCSButton {
        return descendantOfNode(screen, indexes: 2, 3) as JCSButton
    }

    func eventTriggered(notification: NSNotification) {
        events.append(notification)
    }
}